rts/motoko-rts/src/gc/mark_compact/bitmap.rs

```diff
@@ -2,55 +2,126 @@ use crate::mem_utils::memzero;
+
+
+Debug assertions guard the forbidden bytes from access, as this area potentially
+physically overlaps with the Motoko dynamic heap.
+
+## The alignment caveat
+
+For this scheme to work, it is essential that the start of the DH is an address that
+is divisible by 32 (implying that `heap_prefix_words % 8 == 0`). Otherwise the
+`/%8`-operation on the DH's starting address will not yield the least significant bit
+in the BM, and thus the sweep operation will be off.
+
+## Example calculation
+
+Assume the DH is at 0x80000. Assume heap limit being at 0xB0000. Then the BM thus
+could be placed at 0xB0004. Since the heap_prefix_words is 0x20000,
+BITMAP_FORBIDDEN_PTR = 0xB0004 - 0x20000 / 8 = 0xAC004.
+
+Now let's mark the address 0x80548 in the DH. Its absolute word number is 0x20152.
+The `(0x20152 / 8, 0x20152 % 8)`-rule gives a bit position 2 with byte offset 0x402A,
+thus we mark bit 2 in byte 0xAC004 + 0x402A = 0xB002E, which is physically in the BM.
+
+ */
+
 /// Current bitmap
+static mut BITMAP_FORBIDDEN_PTR: *mut u8 = core::ptr::null_mut();
 static mut BITMAP_PTR: *mut u8 = core::ptr::null_mut();
+static mut BITMAP_SIZE: u32 = 0;
 
-pub unsafe fn alloc_bitmap<M: Memory>(mem: &mut M, heap_size: Bytes<u32>) {
+#[cfg(debug_assertions)]
+unsafe fn get_bitmap_forbidden_size() -> usize {
+    BITMAP_PTR as usize - BITMAP_FORBIDDEN_PTR as usize
+}
+
+pub unsafe fn alloc_bitmap<M: Memory>(mem: &mut M, heap_size: Bytes<u32>, heap_prefix_words: u32) {
+    // See Note "How the Wasm-heap maps to the bitmap" above
+    debug_assert_eq!(heap_prefix_words % 8, 0);
     // We will have at most this many objects in the heap, each requiring a bit
     let n_bits = heap_size.to_words().as_u32();
     // Each byte will hold 8 bits.
-    let bitmap_bytes = (n_bits + 7) / 8;
+    BITMAP_SIZE = (n_bits + 7) / 8;
     // Also round allocation up to 8-bytes to make iteration efficient. We want to be able to read
     // 64 bits in a single read and check as many bits as possible with a single `word != 0`.
-    let bitmap_bytes = Bytes(((bitmap_bytes + 7) / 8) * 8);
+    let bitmap_bytes = Bytes(((BITMAP_SIZE + 7) / 8) * 8);
     // Allocating an actual object here as otherwise dump_heap gets confused
     let blob = alloc_blob(mem, bitmap_bytes).get_ptr() as *mut Blob;
     memzero(blob.payload_addr() as usize, bitmap_bytes.to_words());
 
-    BITMAP_PTR = blob.payload_addr()
+    BITMAP_PTR = blob.payload_addr();
+    BITMAP_FORBIDDEN_PTR = BITMAP_PTR.sub(heap_prefix_words as usize / 8)
 }
 
 pub unsafe fn free_bitmap() {
     BITMAP_PTR = core::ptr::null_mut();
+    BITMAP_FORBIDDEN_PTR = core::ptr::null_mut();
+    BITMAP_SIZE = 0;
 }
 
 pub unsafe fn get_bit(idx: u32) -> bool {
-    let byte_idx = idx / 8;
-    let byte = *BITMAP_PTR.add(byte_idx as usize);
-    let bit_idx = idx % 8;
-    (byte >> bit_idx) & 0b1 == 0b1
+    let (byte_idx, bit_idx) = (idx / 8, idx % 8);
+    #[cfg(debug_assertions)]
```

Comment:
```suggestion
```

`debug_assert` already has this guard, it only runs in debug builds.

Reply:
I had compile errors when I removed this.

Reply:
You need to remove `debug_assertions` guard from `get_bitmap_forbidden_size` too.

Expansion of `debug_assert` is something like

```rust
if cfg!(debug_assertions) {
    ... assertion ...
}
```

The condition is compiled to `true` or `false`, but the body is still type checked, so `get_bitmap_forbidden_size` should be defined.

Reply:
I imagine. I cleaned this up in f37b9af30098496ae6edd8c541a9f8c76d59c6b4.